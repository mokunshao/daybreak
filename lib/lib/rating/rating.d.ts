import React from 'react';
import './rating.scss';
export declare const baseClass: (...names: (string | undefined)[]) => string;
interface Props {
    className?: string;
    max?: number;
    value: number;
    onChange: (value: number) => void;
    clearable?: boolean;
}
declare const Rating: React.FC<Props>;
export default Rating;
//# sourceMappingURL=rating.d.ts.map