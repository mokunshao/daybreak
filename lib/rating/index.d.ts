import Rating from './rating';
export default Rating;
export { Rating };
//# sourceMappingURL=index.d.ts.map