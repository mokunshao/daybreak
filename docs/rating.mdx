---
title: Rating
---

import {useState} from 'react';
import { Playground, Props } from 'docz';
import Rating from '../src/rating/rating.tsx';

# Icon

## Basic usage

<Playground>
{() => {
  const [value, setValue] = useState(1);
  return (
    <div>
      <Rating value={value} max={5} onChange={(val) => setValue(val)} />
    </div>
  );
}}
</Playground>

## Clearable

When you repeatedly click on a star, all stars will lose their light

<Playground>
{() => {
  const [value, setValue] = useState(3);
  return (
    <div>
      <Rating clearable value={value} max={6} onChange={(val) => setValue(val)} />
    </div>
  );
}}
</Playground>

## Props

<Props of={Rating} />
