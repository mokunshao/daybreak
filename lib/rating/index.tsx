import Rating from './rating';

export default Rating;
export { Rating };
