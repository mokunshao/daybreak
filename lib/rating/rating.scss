@import '../common.scss';

.daybreak-rating {
  display: inline-flex;

  &-star {
    cursor: pointer;
    font-size: 1.5em;
    color: #ababab;
    user-select: none;

    &-lighten {
      color: #ffd700;
    }
  }
}
