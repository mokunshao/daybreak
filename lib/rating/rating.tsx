import React, { HtmlHTMLAttributes, HTMLAttributes } from 'react';
import './rating.scss';
import { joinedClass } from '../utils/joinedClass';
import { classes } from '../utils/classes';
import Star from './star';

export const baseClass = joinedClass('rating');

interface Props extends HTMLAttributes<HTMLElement>{
  max: number;
  value: number;
}

const Rating: React.FC<Props> = (props) => {
  const {
    className, max, ...rest
  } = props;

  const stars = [];

  for (let index = 0; index < max; index += 1) {
    stars.push(<Star index={index} key={index} />);
  }

  return (
    <div className={classes(baseClass(), className)} {...rest}>
      {stars}
    </div>
  );
};

export default Rating;
