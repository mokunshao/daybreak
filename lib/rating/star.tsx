import React, { useCallback, useState } from 'react';
import { baseClass } from './rating';
import classes from '../utils/classes';

interface StarProps{
  lighten: boolean;
  index: number;
  setOverride: (value: number|null) => void;
  setRating: (value: number|null) => void;
}

const Star: React.FC<StarProps> = (props) => {
  const {
    index, lighten, setOverride, setRating,
  } = props;

  const click = useCallback(() => {
    setRating(index);
  }, [index, setRating]);

  const mouseEnter = useCallback(() => {
    setOverride(index);
  }, [index, setOverride]);

  const mouseLeave = useCallback(() => {
    setOverride(null);
  }, [setOverride]);

  return (
    <span
      className={
        classes(
          baseClass('star'),
          lighten ? baseClass('star', 'lighten') : '',
        )
      }
      onClick={click}
      onMouseEnter={mouseEnter}
      onMouseLeave={mouseLeave}
    >
      â˜…
    </span>
  );
};

export default Star;
