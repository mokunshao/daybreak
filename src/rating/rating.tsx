import React, { useState, HTMLProps, useMemo } from 'react';
import './rating.scss';
import { joinedClass } from '../utils/joinedClass';
import { classes } from '../utils/classes';
import Star from './star';

export const baseClass = joinedClass('rating');

export interface Props extends Omit<HTMLProps<HTMLDivElement>, 'onChange'> {
  max?: number;
  value: number;
  onChange: (value: number) => void;
  clearable?: boolean;
}

export const Rating: React.FC<Props> = React.memo((props) => {
  const {
    className, max = 5, value, onChange, clearable = false, ...rest
  } = props;

  const [override, setOverride] = useState(0);

  const stars = useMemo(() => {
    const array = [];
    for (let index = 1; index <= max; index += 1) {
      array.push(<Star
        lighten={index <= (override || value)}
        index={index}
        key={index}
        rating={value}
        setOverride={setOverride}
        setRating={onChange}
        clearable={clearable}
      />);
    }
    return array;
  }, [clearable, max, onChange, override, value]);

  return (
    <div className={classes(baseClass(), className)} {...rest}>
      {stars}
    </div>
  );
});

export default Rating;
