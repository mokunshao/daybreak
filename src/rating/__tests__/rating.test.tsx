import React, { useState } from 'react';
import { render } from '@testing-library/react';
import { Rating } from '../rating';
import { Star } from '../star';
import Index from '../index';

function Example() {
  const [value, setValue] = useState(1);
  return (
    <div>
      <Rating value={value} max={5} onChange={(val) => setValue(val)} />
    </div>
  );
}

describe('Rating', () => {
  it('exits', () => {
    expect(Rating).toBeTruthy();
    expect(Star).toBeTruthy();
    expect(Index).toBeTruthy();
    expect(<Example />).toBeTruthy();
  });

  it('match snapshot', () => {
    const { asFragment } = render(<Example />);
    expect(asFragment()).toMatchSnapshot();
  });
});
