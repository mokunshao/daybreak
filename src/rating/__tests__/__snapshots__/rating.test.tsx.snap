// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="daybreak-rating"
    >
      <span
        class="daybreak-rating-star daybreak-rating-star-lighten"
        style="transform: scale(1);"
      >
        ★
      </span>
      <span
        class="daybreak-rating-star"
        style="transform: scale(1);"
      >
        ★
      </span>
      <span
        class="daybreak-rating-star"
        style="transform: scale(1);"
      >
        ★
      </span>
      <span
        class="daybreak-rating-star"
        style="transform: scale(1);"
      >
        ★
      </span>
      <span
        class="daybreak-rating-star"
        style="transform: scale(1);"
      >
        ★
      </span>
    </div>
  </div>
</DocumentFragment>
`;
